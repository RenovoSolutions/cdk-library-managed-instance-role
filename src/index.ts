import { aws_iam as iam } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface ManagedInstanceRoleProps {
  /**
   * The managed policies to apply to the role in addition to the default policies.
   */
  readonly managedPolicies?: iam.ManagedPolicy[];
  /**
   * Should the role include directory service access with SSM.
   */
  readonly domainJoinEnabled?: boolean;
  /**
   * Should the role include SSM management. By default if domainJoinEnabled is true then this role is always included.
   */
  readonly ssmManagementEnabled?: boolean;
}

export class ManagedInstanceRole extends Construct {

  public readonly instanceProfile: iam.CfnInstanceProfile;

  constructor(scope: Construct, id: string, props: ManagedInstanceRoleProps) {
    super(scope, id);

    var managedPolicies:iam.IManagedPolicy[] = props.managedPolicies === undefined ? [] : props.managedPolicies;

    let domainJoinEnabled: boolean = props.domainJoinEnabled === undefined ||
                              props.domainJoinEnabled === true;
    let ssmManagementEnabled: boolean = props.ssmManagementEnabled === undefined ||
                                 props.domainJoinEnabled === undefined ||
                                 props.domainJoinEnabled === true ||
                                 props.ssmManagementEnabled === true;
    if (ssmManagementEnabled) { managedPolicies.push(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore')); };
    if (domainJoinEnabled) { managedPolicies.push(iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMDirectoryServiceAccess')); };

    const role = new iam.Role(this, 'role', {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
      managedPolicies,
    });

    this.instanceProfile = new iam.CfnInstanceProfile(this, 'instanceProfile', {
      roles: [role.roleName],
    });
  }
}
