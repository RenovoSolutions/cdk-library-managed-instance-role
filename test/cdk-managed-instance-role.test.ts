import { Stack, App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { ManagedInstanceRole } from '../src/index';

test('Snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack', {});

  new ManagedInstanceRole(stack, 'role', {});

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test('Simple test', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  new ManagedInstanceRole(stack, 'ManagedInstanceRole', {});

  Template.fromStack(stack).resourceCountIs('AWS::IAM::Role', 1);
  Template.fromStack(stack).resourceCountIs('AWS::IAM::InstanceProfile', 1);
});