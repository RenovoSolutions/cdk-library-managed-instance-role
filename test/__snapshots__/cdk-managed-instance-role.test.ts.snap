// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "role08402BCB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMManagedInstanceCore",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonSSMDirectoryServiceAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "roleinstanceProfileFF411810": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "role08402BCB",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
  },
}
`;
